### PR TITLE
Add spring security

### PR DESCRIPTION
**Pull request by igorparada**
- Add spring security
- Add /login page with validation
- Replace unnecessary service interfaces with classes

### CODE REVIEW

File: MVCapplication/src/main/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+application.restServiceURL=http://localhost:8080/rest
```
Comment (kucherif) [inline MVCapplication/src/main/resources/application.properties]: To be fully secured. Could you please add to your backlog to use https with self signed certificate
Maybe this link will help
https://hutter.io/2016/02/09/java-create-self-signed-ssl-certificates-for-tomcat/

File: MVCapplication/src/main/webapp/WEB-INF/views/login.jsp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                <div class="panel-heading">
+                    <h3 class="panel-title">Please Sign In</h3>
+                </div>
+                <div class="panel-body">
+                    <form:form role="form" method="POST" modelAttribute="loginForm">
+                        <fieldset>
+                            <c:if test="${not empty loginError}">
+                                <div class="alert alert-danger">
+                                    <span class="glyphicon glyphicon-minus-sign" aria-hidden="true"></span>
+                                    <c:out value="${loginError}" />
```
Comment (kucherif) [inline MVCapplication/src/main/webapp/WEB-INF/views/login.jsp]: c:out is not required unless you need to display html tags. You can simply use EL ${loginError}

File: RESTserver/src/main/java/com/softserve/if072/restservice/controller/LoginController.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package com.softserve.if072.restservice.controller;
+
+import com.softserve.if072.common.model.User;
+import com.softserve.if072.restservice.dao.mybatisdao.UserDAO;
+import com.softserve.if072.restservice.service.TokenService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.web.bind.annotation.*;
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/controller/LoginController.java]: specify imports

File: RESTserver/src/main/java/com/softserve/if072/restservice/controller/LoginController.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    @RequestMapping(value = "/", method = RequestMethod.POST)
+    @ResponseBody
+    @ResponseStatus(value = HttpStatus.OK)
+    public ResponseEntity<?> getAuthenticationToken(@RequestParam String login, @RequestParam String password, HttpServletResponse response) {
+        try {
+            User user = userDAO.getByUsername(login);
+            if (user == null) {
+                throw new BadCredentialsException("Invalid login");
+            }
+            if (user.getPassword().equals(password) == false) {
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/controller/LoginController.java]: Password in the DB should be stored encrypted.

And this looks weird. This is much better:
`if (!user.getPassword().equals(password)) {`

File: RESTserver/src/main/java/com/softserve/if072/restservice/security/handlers/CustomAuthenticationFailureHandler.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class CustomAuthenticationFailureHandler implements AuthenticationFailureHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(CustomAuthenticationFailureHandler.class);
+
+    @Override
+    public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response, AuthenticationException e) throws IOException, ServletException {
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/security/handlers/CustomAuthenticationFailureHandler.java]: please follow 120 line width rule

File: RESTserver/src/main/java/com/softserve/if072/restservice/service/ShoppingListService.java [hunk 6]
```diff
@@ -1,22 +1,51 @@
...
-    void delete(int id) throws DataNotFoundException;
+@Service
+public class ShoppingListService {
+
+    @Autowired
+    private ShoppingListDAO shoppingListDAO;
+
+    public List<ShoppingList> getByUserId(int user_id) throws DataNotFoundException {
+        List<ShoppingList> list = shoppingListDAO.getByUserID(user_id);
+        if (list != null && !list.isEmpty()){
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/service/ShoppingListService.java]: Please add lib Apache Commons and use CollectionUtils.isNotEmpty() method

File: RESTserver/src/main/java/com/softserve/if072/restservice/service/TokenService.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    public String generateTokenFor(String username) {
+        long expirationDate = DateTime.now().plusSeconds(TOKEN_VALIDITY_TIME).getMillis();
+        String tokenString = buildToken(username, expirationDate);
+        return new String(Base64.encodeBase64(tokenString.getBytes()));
+    }
+
+    private String[] decodeToken(String tokenString) {
+        if (tokenString == null || tokenString.equals("") || !Base64.isBase64(tokenString)) {
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/service/TokenService.java]: StringUtils.isNotBlank() can be used from Apache Commons library
https://commons.apache.org/proper/commons-lang/

File: common-modules/src/main/java/com/softserve/if072/common/model/User.java [hunk 8]
```diff
@@ -11,15 +15,23 @@
 
     private int id;
     private String name;
+
+    @NotEmpty(message = "{error.email.notnull}")
+    @Size(min = 5, max = 64, message = "{error.email.size}")
     private String email;
+
+    @NotEmpty(message = "{error.password.notnull}")
```
Comment (kucherif) [inline common-modules/src/main/java/com/softserve/if072/common/model/User.java]: If you are going to use Hibernate validator, better to use @NotBlank, it consider "   " as empty.

File: RESTserver/src/main/java/com/softserve/if072/restservice/controller/SecurityTestController.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+@RequestMapping(value = "/api")
+public class SecurityTestController {
+
+    @Autowired
+    private UnitDAO unitDAO;
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/controller/SecurityTestController.java]: looks like it is not used.

**Comment by kucherif** [comment]
👍

**Comment by igorparada** [comment]
Thanks for review! I've already fixed the code

**Pull request merged**